File: packages/sdk-ios/Sources/EverframeReporterUI/CompanionPinPresenter.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Built-in attach-PIN surface (spec 2026-08-19). A dedicated UIWindow one
// level above the reporter's (.alert + 2) hosting a passive label — no
// buttons: reading the code to the dashboard IS the consent. Installed by
// `EFReporterPresenter.installResolver()` on iOS and callable directly on
// tvOS hosts (where `installResolver` does not exist).
//
// Gated `#if canImport(UIKit)` ONLY — no `!os(tvOS)`. Apple TV is a PRIMARY
// consumer of companion (see `RelayWSClient`'s header), so this file must
// build and run there, unlike `EFReporterPresenter` (which is iOS-only: TV
// apps route reporting through the phone-companion flow instead, not this
// modal reporter).
//
// Runtime suppression (spec 2026-08-19, controller ruling): `install()`
// itself is unconditional — it always sets `__builtinPinUiInstalled = true`
// so the announce-time capability check degrades honestly. Suppression is
// checked on every `present(_:)` call instead, via
// `CompanionAPI.__builtinPinUiSuppressed` — set by the RN bridge in a later
// task when a custom PIN UI takes over at runtime. `attachChallenge` itself
// keeps updating either way; only the built-in window's rendering is
// silenced, so a custom UI reading the same property never misses a value.
#if canImport(UIKit)
import UIKit
import Combine
import EverframeKit

@MainActor
public enum CompanionPinPresenter {
    private static var window: UIWindow?
    private static var bag = Set<AnyCancellable>()
    private static var installed = false
    private static var expiryTask: Task<Void, Never>?
    private static var activationObserver: NSObjectProtocol?
    // Round-2 review finding 3: `present(_:)` used to require a
    // `.foregroundActive` scene and simply give up otherwise — a challenge
    // that arrived while the app was `.foregroundInactive` (e.g. mid app-
    // switcher transition, a system alert up, or — on tvOS — momentarily
    // losing focus) was NEVER shown, even once the scene came back active,
    // because nothing re-drove `present`. Track the live challenge + its
    // absolute deadline here so `didBecomeActiveNotification` can re-attempt
    // presentation with the REMAINING time.
    private static var currentChallenge: CompanionAttachChallenge?
    private static var currentDeadline: Date?

    /// One-shot installer. Wires a Combine subscription to
    /// `CompanionAPI.attachChallenge` that presents/dismisses the built-in PIN
    /// window, and marks `CompanionAPI.__builtinPinUiInstalled` so the
    /// announce-time capability check can advertise `supportsAttachPin` in
    /// `.builtin` mode. Unconditional — never gated on suppression, which is
    /// a runtime, per-`present` check (see file header).
    ///
    /// Also installs a ONE-TIME `didBecomeActiveNotification` observer (finding
    /// 3): on activation, if a challenge is still live and no window is
    /// currently up, re-present it with the time remaining until its deadline
    /// — never the original `ttlMs`, which would silently extend the code's
    /// lifetime past what the server actually granted.
    public static func install() {
        guard !installed else { return }
        installed = true
        let companion = Everframe.shared.companion
        companion.__builtinPinUiInstalled = true
        companion.$attachChallenge
            .receive(on: DispatchQueue.main)
            .sink { challenge in
                if let challenge { present(challenge) } else { dismiss() }
            }
            .store(in: &bag)
        activationObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: nil
        ) { _ in
            Task { @MainActor in representOnActivationIfNeeded() }
        }
    }

    private static func present(_ challenge: CompanionAttachChallenge) {
        // Suppressed at runtime (RN bridge owns its own PIN UI) — the
        // challenge stays published on `CompanionAPI` either way (a custom UI
        // reads it directly); only the built-in window is silenced. `dismiss`
        // still runs so a window shown before suppression flipped on doesn't
        // linger.
        guard !Everframe.shared.companion.__builtinPinUiSuppressed else {
            dismiss()
            return
        }
        dismiss()
        currentChallenge = challenge
        let deadline = Date().addingTimeInterval(TimeInterval(challenge.ttlMs) / 1000)
        currentDeadline = deadline
        armExpiryTask(deadline: deadline)
        presentWindowIfPossible(challenge: challenge)
    }

    /// (c) The expiry task always arms with the time remaining until the
    /// absolute deadline, never a fresh full `ttlMs` — so re-presenting on
    /// activation (below) can never grant more total lifetime than the
    /// server's original challenge carried.
    private static func armExpiryTask(deadline: Date) {
        expiryTask?.cancel()
        let remaining = CompanionAttachChallenge.remainingMs(deadline: deadline)
        guard remaining > 0 else {
            dismiss()
            return
        }
        expiryTask = Task { @MainActor in
            try? await Task.sleep(nanoseconds: UInt64(remaining) * 1_000_000)
            guard !Task.isCancelled else { return }
            dismiss()
        }
    }

    /// Attempts to put the window up. No-ops (leaves `currentChallenge`/
    /// `currentDeadline` tracked) if no `.foregroundActive` scene exists yet —
    /// `didBecomeActiveNotification` is what re-drives this once one does.
    private static func presentWindowIfPossible(challenge: CompanionAttachChallenge) {
        guard let scene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first(where: { $0.activationState == .foregroundActive })
        else { return }
        let w = UIWindow(windowScene: scene)
        w.windowLevel = .alert + 2
        w.isUserInteractionEnabled = false   // passive — never steals focus/keyboard
        w.rootViewController = CompanionPinViewController(challenge: challenge)
        w.isHidden = false                   // visible WITHOUT becoming key
        window = w
    }

    /// (b) Fires on every `didBecomeActiveNotification`. Only acts when a
    /// challenge is still tracked AND no window is already up (an active
    /// window means either presentation already succeeded, or there's
    /// genuinely nothing to show). Dismisses immediately, without ever
    /// showing a window, if the deadline has already passed while inactive.
    private static func representOnActivationIfNeeded() {
        guard window == nil,
              let challenge = currentChallenge,
              let deadline = currentDeadline
        else { return }
        let remaining = CompanionAttachChallenge.remainingMs(deadline: deadline)
        guard remaining > 0 else {
            dismiss()
            return
        }
        armExpiryTask(deadline: deadline)
        presentWindowIfPossible(challenge: challenge)
    }

    private static func dismiss() {
        expiryTask?.cancel()
        expiryTask = nil
        currentChallenge = nil
        currentDeadline = nil
        window?.isHidden = true
        window?.windowScene = nil
        window = nil
    }
}

/// Passive, top-trailing card: requester name, the PIN in a large monospaced
/// font, and a one-line instruction. No interactive elements — this view
/// controller's window has `isUserInteractionEnabled = false` and is never
/// made key, so it draws only and never fights the host for focus.
@MainActor
final class CompanionPinViewController: UIViewController {
    private let challenge: CompanionAttachChallenge

    init(challenge: CompanionAttachChallenge) {
        self.challenge = challenge
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    override func loadView() {
        // A transparent full-screen container — the visible card is the
        // rounded background view laid out inside it, top-trailing.
        let container = UIView()
        container.backgroundColor = .clear
        view = container
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let card = UIView()
        card.backgroundColor = UIColor.black.withAlphaComponent(0.85)
        card.layer.cornerRadius = 16
        card.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(card)

        let requesterLabel = UILabel()
        requesterLabel.text = "\(challenge.requestedByName) wants to connect"
        requesterLabel.textColor = .white
        requesterLabel.font = .systemFont(ofSize: 18, weight: .semibold)
        requesterLabel.textAlignment = .center
        requesterLabel.numberOfLines = 0

        let codeLabel = UILabel()
        // SECURITY: this is the ONE place the code is ever rendered — never
        // logged, never persisted, never sent anywhere but drawn on screen.
        codeLabel.text = challenge.code
        codeLabel.textColor = .white
        codeLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 40, weight: .bold)
        codeLabel.textAlignment = .center

        let instructionLabel = UILabel()
        instructionLabel.text = "Enter this code in the Everframe dashboard"
        instructionLabel.textColor = UIColor.white.withAlphaComponent(0.8)
        instructionLabel.font = .systemFont(ofSize: 14, weight: .regular)
        instructionLabel.textAlignment = .center
        instructionLabel.numberOfLines = 0

        let stack = UIStackView(arrangedSubviews: [requesterLabel, codeLabel, instructionLabel])
        stack.axis = .vertical
        stack.alignment = .center
        stack.spacing = 8
        stack.translatesAutoresizingMaskIntoConstraints = false
        card.addSubview(stack)

        NSLayoutConstraint.activate([
            card.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 24),
            card.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -24),
            card.widthAnchor.constraint(lessThanOrEqualToConstant: 320),

            stack.topAnchor.constraint(equalTo: card.topAnchor, constant: 20),
            stack.bottomAnchor.constraint(equalTo: card.bottomAnchor, constant: -20),
            stack.leadingAnchor.constraint(equalTo: card.leadingAnchor, constant: 24),
            stack.trailingAnchor.constraint(equalTo: card.trailingAnchor, constant: -24),
        ])
    }
}
#endif

File: packages/sdk-android/android/everframe-reporter-ui/src/main/kotlin/dev/everframe/ui/CompanionPinPresenter.kt
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Built-in attach-PIN surface (spec 2026-08-19). Collects
// Companion.attachChallenge and shows a passive dialog over the
// currently-resumed Activity — no buttons; reading the code to the dashboard
// IS the consent. If no Activity is resumed when a challenge lands
// (backgrounded app), the challenge is held and presented on the next
// resume, still gated on the flow's CURRENT value so an expired/cleared
// challenge never pops late.
//
// `:everframe-reporter-ui` declares no `res/values/strings.xml` (checked —
// every other file in this module uses inline literals, not `R.string`), so
// this file follows suit rather than introducing the module's first
// resource file for two short strings.
package dev.everframe.ui

import android.app.Activity
import android.app.AlertDialog
import android.graphics.Typeface
import android.view.Gravity
import android.widget.LinearLayout
import android.widget.TextView
import dev.everframe.Everframe
import dev.everframe.companion.AttachChallengeInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

internal object CompanionPinPresenter {
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var dialog: AlertDialog? = null
    private var installed = false
    private var expiryJob: Job? = null

    // Stale re-present fix (spec 2026-08-19 review finding 2). [onActivityResumed]
    // re-presents from the flow's CURRENT value, which is the SAME
    // [AttachChallengeInfo] instance the collector below already presented —
    // re-arming [present] with a fresh full `challenge.ttlMs` on every resume
    // would let the on-device dialog outlive its real expiry indefinitely
    // (each resume resets the clock). The fix: compute an ABSOLUTE deadline
    // once, the moment a given challenge is first observed, and always arm
    // the expiry job against the REMAINING time to that deadline — never a
    // fresh full ttlMs on a re-present. A single field suffices (not a map):
    // `Companion.attachChallenge` is a `StateFlow<AttachChallengeInfo?>`, so
    // at most one challenge is ever live at a time. Keyed by structural
    // equality of the whole [AttachChallengeInfo] (not just `code`) so a
    // genuinely NEW challenge that happens to reuse the same code (distinct
    // `ttlMs`/`requestedByName`) is correctly treated as new, not stale.
    private var trackedChallenge: AttachChallengeInfo? = null
    private var trackedDeadlineElapsedRealtime: Long = 0L

    fun install() {
        if (installed) return
        installed = true
        Everframe.__attachPinUiInstalled = true
        scope.launch {
            dev.everframe.companion.Companion.attachChallenge.collect { challenge ->
                if (challenge == null) {
                    trackedChallenge = null
                    dismiss()
                } else {
                    if (challenge != trackedChallenge) {
                        // Genuinely new challenge (first sighting) — this is
                        // the ONE place a full `ttlMs` window is minted.
                        trackedChallenge = challenge
                        trackedDeadlineElapsedRealtime =
                            android.os.SystemClock.elapsedRealtime() + challenge.ttlMs
                    }
                    present(challenge, trackedDeadlineElapsedRealtime)
                }
            }
        }
    }

    /** Re-check on foreground: ActivityRegistry calls this from onActivityResumed. */
    fun onActivityResumed() {
        val challenge = dev.everframe.companion.Companion.attachChallenge.value
        if (challenge == null || dialog != null) return
        // The ordinary case: this IS the challenge the collector above
        // already tracked — reuse its recorded deadline so a re-present on
        // resume dismisses at the SAME wall-clock moment the original
        // `present` would have, not a fresh `ttlMs` later.
        val deadline = if (challenge == trackedChallenge) {
            trackedDeadlineElapsedRealtime
        } else {
            // Defensive only: the flow changed underneath us without the
            // collector observing it first (collector and this function both
            // read the same StateFlow, so this should not happen in
            // practice) — treat it as a fresh sighting rather than presenting
            // against a stale/zero deadline.
            trackedChallenge = challenge
            (android.os.SystemClock.elapsedRealtime() + challenge.ttlMs)
                .also { trackedDeadlineElapsedRealtime = it }
        }
        present(challenge, deadline)
    }

    /**
     * An `AlertDialog` leaks if its Activity finishes underneath it.
     * `ActivityRegistry.onActivityPaused`/`onActivityDestroyed` call this
     * when clearing `current` — the challenge stays in the flow, so the next
     * resume ([onActivityResumed]) re-presents it against whatever Activity
     * comes back.
     */
    fun onActivityGone() {
        dismiss()
    }

    private fun present(challenge: AttachChallengeInfo, deadlineElapsedRealtime: Long) {
        // Suppressed by the host (RN bridge / other native hosts) when the
        // configured attach-PIN mode is not `builtin` — something else owns
        // rendering. The StateFlow itself is untouched; a custom UI still
        // collects it directly. `dismiss()` still runs so a dialog shown
        // before suppression flipped on doesn't linger (matches iOS).
        if (Everframe.__attachPinUiSuppressed) {
            dismiss()
            return
        }
        dismiss()
        // Local expiry: the server's cleared/expired frame rides the sweeper
        // (60s tick) — hide at the deadline locally so a dead code never
        // lingers. ALWAYS the remaining time to the tracked absolute
        // deadline, never a fresh full `challenge.ttlMs` — see the class-level
        // comment on `trackedDeadlineElapsedRealtime` for why a re-present on
        // resume must not reset this clock.
        val remaining = deadlineElapsedRealtime - android.os.SystemClock.elapsedRealtime()
        if (remaining <= 0) {
            // Already past its deadline (e.g. resumed long after the original
            // ttlMs window closed) — never show a code that's already dead.
            return
        }
        expiryJob?.cancel()
        expiryJob = scope.launch {
            delay(remaining)
            dismiss()
        }
        val activity: Activity = ActivityRegistry.activeActivity() ?: return // onActivityResumed retries
        val content = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER_HORIZONTAL
            setPadding(48, 40, 48, 40)
            addView(
                TextView(activity).apply {
                    text = "${challenge.requestedByName} wants to connect"
                    textSize = 14f
                },
            )
            addView(
                TextView(activity).apply {
                    // SECURITY: challenge.code is never logged anywhere in this
                    // class — display on-device IS the feature, logcat is not
                    // the device screen.
                    text = challenge.code
                    textSize = 40f
                    typeface = Typeface.MONOSPACE
                    letterSpacing = 0.3f
                },
            )
            addView(
                TextView(activity).apply {
                    text = "Enter this code in the Everframe dashboard"
                    textSize = 12f
                },
            )
        }
        dialog = AlertDialog.Builder(activity)
            .setView(content)
            .setCancelable(false) // dismissal is server-driven (cleared/expiry), not a tap
            .create()
            .also { it.show() }
    }

    private fun dismiss() {
        expiryJob?.cancel()
        expiryJob = null
        dialog?.let { runCatching { it.dismiss() } }
        dialog = null
    }
}
